### PR TITLE
Auto-generate the rules table of contents

### PR DESCRIPTION
**Pull request by phillipuniverse**
Follow-up to https://github.com/charliermarsh/ruff/pull/1109#discussion_r1041740034

### CODE REVIEW

File: ruff_dev/src/generate_rules_table.rs [hunk 1]
```diff
@@ -23,77 +26,90 @@ pub struct Cli {
...
-            .expect("Unable to find end pragma");
-        let suffix = &existing[index..];
-
-        // Write the prefix, new contents, and suffix.
-        let mut f = OpenOptions::new().write(true).truncate(true).open(&file)?;
-        write!(f, "{prefix}\n\n")?;
-        write!(f, "{output}")?;
-        write!(f, "{suffix}")?;
+        let _toc_ok = replace_readme_section(toc_out, TOC_BEGIN_PRAGMA, TOC_END_PRAGMA);
+        let _table_ok = replace_readme_section(table_out, TABLE_BEGIN_PRAGMA, TABLE_END_PRAGMA);
```
Comment (phillipuniverse) [inline ruff_dev/src/generate_rules_table.rs:82]: I wasn't sure what the "right" way to handle multiple results here. I wanted to do something like:

```rust
return Result(toc_ok, table_ok)
```

or pythonic-like:

```python
return toc_ok and table_ok
```

Comment (charliermarsh) [inline ruff_dev/src/generate_rules_table.rs:82]: I think here, you probably want:

```rust
replace_readme_section(toc_out, TOC_BEGIN_PRAGMA, TOC_END_PRAGMA)?;
replace_readme_section(table_out, TABLE_BEGIN_PRAGMA, TABLE_END_PRAGMA)?;
```

That way, if either call returns an `Err`, this caller function will stop immediately and return that error.

(As-written, the code will basically ignore / drop the errors.)

File: ruff_dev/src/generate_rules_table.rs [hunk 2]
```diff
@@ -23,77 +26,90 @@ pub struct Cli {
...
+                table_out.push('\n');
             }
         }
-        output.push('\n');
+        table_out.push('\n');
     }
 
     if cli.dry_run {
-        print!("{output}");
+        print!("Table of Contents: {toc_out}");
```
Comment (phillipuniverse) [inline ruff_dev/src/generate_rules_table.rs:79]: Whoops - fixing this

File: ruff_dev/src/generate_rules_table.rs (hunk 1 above)
Comment (phillipuniverse) [reply ruff_dev/src/generate_rules_table.rs:82]: Got it, thanks!

Comment (charliermarsh) [reply ruff_dev/src/generate_rules_table.rs:82]: See my other comment!

If you're calling a function that returns `Result<T>` from a function that returns `Result<T>` (as is the case here), you can always do `let ok_value = foo()?`, which will leave `ok_value` as type `T` but return immediately if `foo` returned an error.

**Comment by charliermarsh** [comment]
Seeing a few minor lint errors when running `cargo +nightly clippy --workspace --all-targets --all-features -- -W clippy::pedantic` but otherwise good to go:

```
warning: variables can be used directly in the `format!` string
  --> ruff_dev/src/generate_rules_table.rs:33:29
   |
33 |         table_out.push_str(&format!("### {} ({})", check_category.title(), codes_csv));
   |                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#uninlined_format_args
   = note: `-W clippy::uninlined-format-args` implied by `-W clippy::pedantic`
help: change this to
   |
33 -         table_out.push_str(&format!("### {} ({})", check_category.title(), codes_csv));
33 +         table_out.push_str(&format!("### {} ({codes_csv})", check_category.title()));
   |

warning: single-character string constant used as pattern
  --> ruff_dev/src/generate_rules_table.rs:41:59
   |
41 |             check_category.title().to_lowercase().replace(" ", "-"),
   |                                                           ^^^ help: try using a `char` instead: `' '`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#single_char_pattern
   = note: `#[warn(clippy::single_char_pattern)]` on by default

warning: single-character string constant used as pattern
  --> ruff_dev/src/generate_rules_table.rs:42:64
   |
42 |             codes_csv.to_lowercase().replace(",", "-").replace(" ", "")
   |                                                                ^^^ help: try using a `char` instead: `' '`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#single_char_pattern

warning: single-character string constant used as pattern
  --> ruff_dev/src/generate_rules_table.rs:42:46
   |
42 |             codes_csv.to_lowercase().replace(",", "-").replace(" ", "")
   |                                              ^^^ help: try using a `char` instead: `','`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#single_char_pattern

warning: this argument is passed by value, but not consumed in the function body
  --> ruff_dev/src/generate_rules_table.rs:88:36
   |
88 | fn replace_readme_section(content: String, begin_pragma: &str, end_pragma: &str) -> Result<()> {
   |                                    ^^^^^^ help: consider changing the type to: `&str`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_pass_by_value
   = note: `-W clippy::needless-pass-by-value` implied by `-W clippy::pedantic`

warning: using `write!()` with a format string that ends in a single newline
   --> ruff_dev/src/generate_rules_table.rs:110:5
    |
110 |     write!(f, "{prefix}\n")?;
    |     ^^^^^^^^^^^^^^^^^^^^^^^
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#write_with_newline
    = note: `#[warn(clippy::write_with_newline)]` on by default
help: use `writeln!` instead
    |
110 -     write!(f, "{prefix}\n")?;
110 +     writeln!(f, "{prefix}")?;
    |
```

**Comment by charliermarsh** [comment]
Ohh interesting, because it doesn't like the HTML comment?

What about this?

Screen Shot 2022-12-07 at 9 44 54 AM

**Comment by phillipuniverse** [comment]
@charliermarsh great success!

image

**Comment by charliermarsh** [comment]
Awesome, thank you!

**Pull request merged**